### PR TITLE
[ServiceNow] Documentation for ServiceNow connector

### LINKED ISSUE #898: ServiceNow: Documentation

### PR DESCRIPTION
**Pull request by parth-elastic**
#### Pre-Review Checklist
- [x] Considered corresponding documentation changes

### CODE REVIEW

File: docs/DEVELOPING.md [hunk 1]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+ServiceNow "Tokyo", "San Diego" & "Rome" versions are compatible with Elastic connector frameworks.
+
+### Configuration
+
+The following configuration fields need to be provided for setting up the connector:
+
+##### `url`
+
+The host url of the account for ServiceNow.
```
Comment (timgrein) [inline docs/DEVELOPING.md:470]:
```suggestion
The host url of the ServiceNow instance.
```

We should probably refer to the instance and not the account for the host url, right?

File: docs/DEVELOPING.md [hunk 2]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+##### `username`
+
+The username of the account for ServiceNow.
+
+##### `password`
+
+The password of the account to be used for the ServiceNow.
```
Comment (timgrein) [inline docs/DEVELOPING.md:478]:
```suggestion
The password of the account used for ServiceNow.
```

File: docs/DEVELOPING.md [hunk 3]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+##### `services`
+
+Comma-separated list of Services to fetch data from ServiceNow. If the value is `*`, the connector will fetch data from list of basic services provided by ServiceNow: `Users`, `Incidents`, `Request Items`, `Knowledge Articles`, `Change Requests`. Default value is `*`. Examples:
```
Comment (timgrein) [inline docs/DEVELOPING.md:482]:
```suggestion
Comma-separated list of services to fetch data from ServiceNow. If the value is `*`, the connector will fetch data from the list of basic services provided by ServiceNow: `Users`, `Incidents`, `Request Items`, `Knowledge Articles`, `Change Requests`. Default value is `*`. Examples:
```

File: docs/DEVELOPING.md [hunk 4]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+  - `Incident, User`
+  - `*`
+
+##### `retry_count`
+
+The number of retry attempts after failed request to ServiceNow. Default value is `3`.
```
Comment (timgrein) [inline docs/DEVELOPING.md:489]:
```suggestion
The number of retry attempts after a failed request to ServiceNow. Default value is `3`.
```

File: docs/DEVELOPING.md [hunk 5]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+The connector syncs the following ServiceNow object types: 
+- **Records**
+- **Attachments**
+
+### Sync rules
+
+- Content of files bigger than 10 MB won't be extracted.
+- Permissions are not synced. **All documents** indexed to an Elastic deployment will be visible to **all users with access** to that Elastic Deployment.
+- Filtering rules are not available in the present version. Currently filtering is controlled via ingest pipelines.
```
Comment (timgrein) [inline docs/DEVELOPING.md:505]:
```suggestion
- Advanced sync rules are not available in the present version. Currently filtering is controlled via ingest pipelines.
```

Basic rules are available to every connector from a technical point of view. cc @danajuratoni Do we want to enable basic sync rules for ServiceNow or is there a good reason to disable them?

**Review by timgrein — COMMENTED** [review]
Only having some minor comments, I would like to wait for another review from docs/support/product before merging it.

**Comment by leemthompo** [comment]
Heads up I won't be able to review this until next week. :)

**Comment by danajuratoni** [comment]
Unrelated but just wanted to point out an inconsistency
https://github.com/elastic/connectors-python/blob/1f14e1316b67e7f952ae48f8b3cf242c748d2a23/docs/DEVELOPING.md?plain=1#L284

>This connector is in **beta** and is subject to change.

@leemthompo we decided to release SharePoint in Tech preview.

File: docs/DEVELOPING.md [hunk 6]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+##### `password`
+
+The password of the account used for ServiceNow.
+
+##### `services`
+
+Comma-separated list of services to fetch data from ServiceNow. If the value is `*`, the connector will fetch data from the list of basic services provided by ServiceNow: `Users`, `Incidents`, `Request Items`, `Knowledge Articles`, `Change Requests`. Default value is `*`. Examples:
+
+  - `Incident, User`
```
Comment (danajuratoni) [inline docs/DEVELOPING.md:484]: The description above mentions `Incidents`, the example `Incident`. Which one is correct?

File: docs/DEVELOPING.md [hunk 7]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+To use this connector as a **connector client**, use the **Customized connector** workflow.
+
+For additional operations, see [Usage](https://www.elastic.co/guide/en/enterprise-search/master/connectors-usage.html).
+
+### Compatibility
+
+ServiceNow "Tokyo", "San Diego" & "Rome" versions are compatible with Elastic connector frameworks.
+
+### Configuration
```
Comment (danajuratoni) [inline docs/DEVELOPING.md:464]: @leemthompo I think 8.8 is a good release to review the Configuration section.

On one hand we have **native connectors** - where users should not care about the parameter name / how the configuration field is called in the config.yml.
On the other hand we have **connector clients** - where users might take advantage of advanced config options (such as retry_count / concurrent_downloads) which are not shown is UI by default.

Also, the labels are often different from the parameter used internally.

@navarone-feekery do you have RCF documentation we could use as a reference when reviewing this?

File: docs/DEVELOPING.md (hunk 6 above)
Comment (parth-elastic) [reply docs/DEVELOPING.md:484]: Updated basic services list as per ServiceNow's terminology 👍

File: docs/DEVELOPING.md (hunk 7 above)
Comment (navarone-feekery) [reply docs/DEVELOPING.md:464]: @danajuratoni is the [connector protocol](https://github.com/elastic/connectors-python/blob/main/docs/CONNECTOR_PROTOCOL.md) sufficient? I added the new RCF to it. It doesn't contain an example, though.

Comment (leemthompo) [reply docs/DEVELOPING.md:464]: @danajuratoni >@leemthompo I think 8.8 is a good release to review the Configuration section.

TBC 8.8 or 8.9?

**Comment by parth-elastic** [comment]
Hey @leemthompo @ppf2 , Did you get a chance to review this PR? Just the documentation PR is remaining for the ServiceNow connector. It would be great if you guys can review this PR as this PR is open from quite a long time.

cc: @danajuratoni

**Comment by parth-elastic** [comment]
Thanks @leemthompo for the quick review. Updated PR as per requested changes.

@ppf2 It would be great if you also can take look and share your feedback for the same.

File: docs/DEVELOPING.md [hunk 8]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
 
 ℹ️ The connector uses the Elastic [ingest attachment processor](https://www.elastic.co/guide/en/enterprise-search/current/ingest-pipelines.html) plugin for extracting file contents. The ingest attachment processor extracts files by using the Apache text extraction library Tika. Supported file types eligible for extraction can be found as `TIKA_SUPPORTED_FILETYPES` in [utils.py](../connectors/utils.py) file.
 
+## ServiceNow Connector
+
+The [Elastic ServiceNow connector](../connectors/sources/servicenow.py) is provided in the Elastic connectors python framework and can be used as a [connector client](https://www.elastic.co/guide/en/enterprise-search/current/build-connector.html).
+
+### Availability and prerequisites
```
Comment (ppf2) [inline docs/DEVELOPING.md:448]: It seems like there should be some information on what user and role is required to query the services.

File: docs/DEVELOPING.md [hunk 9]
```diff
@@ -441,6 +441,97 @@ $ make ftest NAME=sharepoint
...
+
+The username of the account for ServiceNow.
+
+##### `password`
+
+The password of the account used for ServiceNow.
+
+##### `services`
+
+Comma-separated list of services to fetch data from ServiceNow. If the value is `*`, the connector will fetch data from the list of basic services provided by ServiceNow: `User`, `Incident`, `Requested Item`, `Knowledge`, `Change Request`. Default value is `*`. Examples:
```
Comment (ppf2) [inline docs/DEVELOPING.md:482]: Are we using ServiceNow terminology here?  I can't find any information online on what the full list of "basic services" available in ServiceNow are.  Can we link "services" to the appropriate documentation in ServiceNow so it's more clear?

Comment (parth-elastic) [reply docs/DEVELOPING.md]: Yes, these are ServiceNow terminologies.
There is nothing like basic service. However as there are thousands of services are available in ServiceNow but we are fetching only 5 frequently used service referred from Workplace search.

Also, to make things more clear we've added links for the respective service

File: docs/DEVELOPING.md [hunk 10]
```diff
@@ -479,7 +481,7 @@ The password of the account used for ServiceNow.
...
 
-Comma-separated list of services to fetch data from ServiceNow. If the value is `*`, the connector will fetch data from the list of basic services provided by ServiceNow: `User`, `Incident`, `Requested Item`, `Knowledge`, `Change Request`. Default value is `*`. Examples:
+Comma-separated list of services to fetch data from ServiceNow. If the value is `*`, the connector will fetch data from the list of basic services provided by ServiceNow: [Users](https://docs.servicenow.com/bundle/utah-platform-administration/page/administer/roles/concept/user.html), [Incidents](https://docs.servicenow.com/bundle/tokyo-it-service-management/page/product/incident-management/concept/c_IncidentManagement.html), [Request Items](https://docs.servicenow.com/bundle/tokyo-servicenow-platform/page/use/service-catalog-requests/task/t_AddNewRequestItems.html), [Knowledge Articles](https://docs.servicenow.com/bundle/tokyo-customer-service-management/page/product/customer-service-management/task/t_SearchTheKnowledgeBase.html), [Change Requests](https://docs.servicenow.com/bundle/tokyo-it-service-management/page/product/change-management/task/t_CreateAChange.html). Default value is `*`. Examples:
```
[the remark this answers is not in the document]
Comment (parth-elastic) [reply docs/DEVELOPING.md]: Done 👍

File: docs/DEVELOPING.md (hunk 6 above)
Comment (danajuratoni) [reply docs/DEVELOPING.md:493]: Please add the entire list in the first example to also show if a separator is expected. e.g. `knowledge-article` or `knowledge_article`

File: docs/DEVELOPING.md [hunk 11]
```diff
@@ -441,6 +441,106 @@ $ make ftest NAME=sharepoint
...
+Default value is `*`. Examples:
+
+  - `Incident, User`
+  - `*`
+
+##### `retry_count`
+
+The number of retry attempts after a failed request to ServiceNow. Default value is `3`.
+
+##### `concurrent_downloads`
```
Comment (danajuratoni) [inline docs/DEVELOPING.md:500]: The TDD does not list this parameter. Is this needed?

File: docs/DEVELOPING.md (hunk 6 above)
Comment (parth-elastic) [reply docs/DEVELOPING.md]: Sure, I've added the entire list. Also user will have to write the entire service name as present in ServiceNow irrespective of separator.

File: docs/DEVELOPING.md [hunk 12]
```diff
@@ -441,6 +441,106 @@ $ make ftest NAME=sharepoint
...
+- [Requested Item](https://docs.servicenow.com/bundle/tokyo-servicenow-platform/page/use/service-catalog-requests/task/t_AddNewRequestItems.html)
+- [Knowledge](https://docs.servicenow.com/bundle/tokyo-customer-service-management/page/product/customer-service-management/task/t_SearchTheKnowledgeBase.html)
+- [Change Request](https://docs.servicenow.com/bundle/tokyo-it-service-management/page/product/change-management/task/t_CreateAChange.html)
+
+Default value is `*`. Examples:
+
+  - `User, Incident, Requested Item, Knowledge, Change Request`
+  - `*`
+
+##### `retry_count`
```
Comment (danajuratoni) [inline docs/DEVELOPING.md:496]: `retry count` and `concurrent downloads` should be advanced config, correct?

File: docs/DEVELOPING.md [hunk 13]
```diff
@@ -441,6 +441,106 @@ $ make ftest NAME=sharepoint
...
+
+The connector syncs the following ServiceNow object types: 
+- **Records**
+- **Attachments**
+
+### Sync rules
+
+- Content of files bigger than 10 MB won't be extracted.
+- Permissions are not synced. **All documents** indexed to an Elastic deployment will be visible to **all users with access** to that Elastic Deployment.
+- Advanced sync rules are not available in the present version. Currently filtering is controlled via ingest pipelines.
```
Comment (danajuratoni) [inline docs/DEVELOPING.md:514]: Sync rules should be part of the 8.9 ServiceNow implementation.

**Comment by danajuratoni** [comment]
LGTM, added a few comments.

File: docs/DEVELOPING.md (hunk 12 above)
Comment (parth-elastic) [reply docs/DEVELOPING.md:496]: Yes, correct they are part of advance config. User will not be able to see these fields in UI.

File: docs/DEVELOPING.md (hunk 13 above)
Comment (parth-elastic) [reply docs/DEVELOPING.md:514]: Yes, Once we raise Sync rules PR will updating documentation accordingly.

**Comment by parth-elastic** [comment]
@danajuratoni @timgrein can you please approve this PR so we can merge this, If everything looks good to you

**Review by timgrein — APPROVED** [review]
LGTM, thank you for the changes!

**Pull request merged**